src/nlp/search.py

```diff
@@ -0,0 +1,366 @@
+        self._indexes[column].add_vectors(vectors, column=column)
+
+    def add_text_index(self, column: str, texts, es_client, index_name):
+        """ Add a text index using ElasticSearch for fast retrieval.
+
+            Args:
+                `column` (`str`): The column/identifier of the index. This is the column that is used to call `.get_nearest` or `.search`.
+                `texts` (`Union[List[str], nlp.Dataset]`): The texts to index. It can be a `nlp.Dataset`.
+                `es_client` (`elasticsearch.Elasticsearch`): The elasticsearch client used to create the index.
+                `index_name` (Optional `str`): The elasticsearch index name used to create the index.
+                `column` (Optional `str`): In the case of a `nlp.Dataset` input, which column to index.
+        """
+        self._indexes[column] = SparseIndex(es_client, index_name)
+        self._indexes[column].add_texts(texts, column=column)
+
+    def drop_index(self, column: str):
+        """ Drop the index with the specified column.
+
+            Args:
+                `column` (`str`): The column/identifier of the index.
+        """
+        del self._indexes[column]
+
+    def search(self, column: str, query, k: int = 10) -> Tuple[List[float], List[int]]:
+        """ Find the nearest examples indices in the dataset to the query.
+
+            Args:
+                `column` (`str`): The name/identifier of the index.
+                `query` (`Union[str, np.ndarray]`): The query as a string if `column` is a text index or as a numpy array if `column` is a vector index.
+                `k` (`int`): The number of examples to retrieve.
+
+            Ouput:
+                `scores` (`List[List[float]`): The retrieval scores of the retrieved examples.
+                `indices` (`List[List[int]]`): The indices of the retrieved examples.
+        """
+        self._check_index_is_initialized(column)
+        return self._indexes[column].search(query, k)
+
+    def search_batch(self, column: str, queries, k: int = 10) -> Tuple[List[List[float]], List[List[int]]]:
+        """ Find the nearest examples indices in the dataset to the query.
+
+            Args:
+                `column` (`str`): The column/identifier of the index.
+                `queries` (`Union[List[str], np.ndarray]`): The queries as a list of strings if `column` is a text index or as a numpy array if `column` is a vector index.
+                `k` (`int`): The number of examples to retrieve per query.
+
+            Ouput:
+                `scores` (`List[List[float]`): The retrieval scores of the retrieved examples per query.
+                `indices` (`List[List[int]]`): The indices of the retrieved examples per query.
+        """
+        self._check_index_is_initialized(column)
+        return self._indexes[column].search_batch(queries, k)
+
+    def get_nearest(self, column: str, query, k: int = 10) -> Tuple[List[float], List[dict]]:
```

Comment:
(nit) `get_nearest_example` is better IMO.

Reply:
not a fan of single letter variables `k` -> why not type checking for `query`?

Reply:
I like `get_nearest_examples ` too.
Let me type check query too.

About the name of `k`, it follows some naming convention widely used (cf 'k nearest neighbors').
Examples:
- sklearn.neighbors.KDTree.query uses `k`
- faiss.IndexFlat.search uses `k`
- annoy. AnnoyIndex.get_nns_by_vector uses `search_k `
- dlib/graphtools find_approximate_k_nearest_neighbors uses `k`

Sometimes it is called something different (`nn` for flann, `final_num_examples` for ScaNN...)